sdk/iot/hub/src/az_iot_hub_client_methods.c

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+
+#include "az_iot_hub_client.h"
+#include <az_precondition_internal.h>
+#include <az_result.h>
+#include <az_span.h>
+
+#include <_az_cfg.h>
+
+static const az_span methods_topic_prefix = AZ_SPAN_LITERAL_FROM_STR("$iothub/methods/");
+static const az_span methods_topic_filter_suffix = AZ_SPAN_LITERAL_FROM_STR("POST/");
+static const az_span methods_response_topic_result = AZ_SPAN_LITERAL_FROM_STR("res/");
+static const az_span methods_response_topic_properties = AZ_SPAN_LITERAL_FROM_STR("/?$rid=");
+static const uint8_t hash_tag = '#';
+
+
+AZ_NODISCARD az_result az_iot_hub_client_methods_subscribe_topic_filter_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_topic_filter,
+    az_span* out_mqtt_topic_filter)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_topic_filter, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic_filter);
+
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, methods_topic_prefix, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, methods_topic_filter_suffix, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append_uint8(mqtt_topic_filter, hash_tag, &mqtt_topic_filter));
+
+  *out_mqtt_topic_filter = mqtt_topic_filter;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_hub_client_methods_received_topic_parse(
+    az_iot_hub_client const* client,
+    az_span received_topic,
+    az_iot_hub_client_method_request* out_request)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(received_topic, 1, false);
+  AZ_PRECONDITION_NOT_NULL(out_request);
+
+  int32_t index = az_span_find(received_topic, methods_topic_prefix);
+
+  if (index == -1) 
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
```

Comment:
I think we should create a return value for IoT. Spitballing here but something like `AZ_ERROR_IOT_PARSE_NOT_FEATURE`, `AZ_ERROR_IOT_NOT_MINE`....

Reply:
This is expected isn't it, for like case where app has one function handle ALL MQTT incoming traffic and then will "visit" each of the method_parse/twin_parse/etc.?  If so we'd need a new IoT code as and it shouldn't be `ERROR` in it.  `AZ_IOT_NOT_PROCESSED`??

Reply:
@jspaith , I think we need to separate the concepts here. `received_topic` could be:
1. A valid direct method request topic;
2. A valid topic name for the other features we support;
3. An invalid topic name.

The function covers 1. because it's the happy path.
But it does not differentiate between 2. and 3. So to say that the return code for these cases should be AZ_IOT_NOT_PROCESSED hides the case where the customer has passed a case 3. input. 
Of course it makes no sense to code separate logics to differentiate between 2. and 3. in the function, it would be duplication of code and expensive.
So I understand that the function should return a *_ERROR_* in case it cannot parse the input. In the upper layer the topic will be submitted to each parsing function and get errors over and over, until one returns OK.
I do agree with Dane though, the return should be an AZ_IOT enum value.
+ @CIPop @hihigupt  to chime in.

Reply:
Yea I don't think there is even a way to differentiate between 2 and 3 since each feature is not aware of the others. I think in that case we should have one of two returns:
1. Parsed and it is mine. Success
2. I don't know what this is but it isn't mine. Error

Reply:
OK - I'm fine with proposal for parsed OK/not mine.  NIT: I'm not convinced though this is an error state -- error implies to me something unexpected that'd be akin to an exception.  This feels more like informational.  I don't have great idea of how to name this though.

RE potentially 3 error codes I'd considered -- you guys are right, not needed.  Even in the case where let's say we get a "/method/POST-thingy/first-partOK" that matches 1st part of topic, and followed by "total-garbage-that-we-can't-parse" I still think we're OK.  We'd return the NOT_MIne, and none of other handlers would process it either and application would still be in state of "I don't know how to process this."  So agree we don't need a special error code for the scenario in this paragraph.

Reply:
So two names possibly here:
`AZ_IOT_INVALID_FEATURE_TOPIC` or
`AZ_ERROR_INVALID_FEATURE_TOPIC`

Just like the core has multiple success cases here:
https://github.com/Azure/azure-sdk-for-c/blob/a86d152733e2dbec9106b230ba283ac6aae282d3/sdk/core/core/inc/az_result.h#L67-L68
We can have our own different IoT successes. Any thoughts?

Reply:
`INVALID` though may be strong?

If I pass a twin topic to this function, it's not topic's fault we passed it in the wrong place whereas invalid makes it sound like a bug in server.  Maybe `no_match` or I was actually even thinking about a `continue` already in this.  (In Windows structured exception handling, if your exception handler doesn't want the exception you mark it with a continue to let the kernel keep unwinding to the next exception handler or else the secondary exception (debugger/crash)).

Reply:
+1 on invalid being strong - it may be valid.
AZ_ERROR_FEATURE_TOPIC_IGNORED + docs on saying this means the topic is not meant for this feature.

Reply:
Ok, using AZ_ERROR_IOT_TOPIC_NO_MATCH